src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs

```diff
@@ -271,109 +250,63 @@ public Task<byte[]> GetByteArrayAsync(Uri? requestUri, CancellationToken cancell
-    #endif
-                    }
+                // Get the response content.
+                HttpContent c = response.Content;
+                if (HttpTelemetry.Log.IsEnabled() && telemetryStarted)
+                {
+                    HttpTelemetry.Log.ResponseContentStart();
+                    responseContentTelemetryStarted = true;
+                }
 
-                    // No content to return.
-                    return Array.Empty<byte>();
+                // If we got a content length, then we assume that it's correct and create a MemoryStream
+                // to which the content will be transferred.  That way, assuming we actually get the exact
+                // amount we were expecting, we can simply return the MemoryStream's underlying buffer.
+                // If we didn't get a content length, then we assume we're going to have to grow
+                // the buffer potentially several times and that it's unlikely the underlying buffer
+                // at the end will be the exact size needed, in which case it's more beneficial to use
+                // ArrayPool buffers and copy out to a new array at the end.
+                long? contentLength = c.Headers.ContentLength;
+                buffer = contentLength.HasValue ?
+                    new HttpContent.LimitMemoryStream(_maxResponseContentBufferSize, (int)contentLength.GetValueOrDefault()) :
+                    new HttpContent.LimitArrayPoolWriteStream(_maxResponseContentBufferSize);
+
+                using Stream responseStream = c.TryReadAsStream() ?? await c.ReadAsStreamAsync(cts.Token).ConfigureAwait(false);
+                try
+                {
+                    await responseStream.CopyToAsync(buffer, cts.Token).ConfigureAwait(false);
                 }
+                catch (Exception e) when (HttpContent.StreamCopyExceptionNeedsWrapping(e))
+                {
+                    throw HttpContent.WrapStreamCopyException(e);
+                }
+
+                return
+                    buffer.Length == 0 ? Array.Empty<byte>() :
+                    buffer is HttpContent.LimitMemoryStream lms ? lms.GetSizedBuffer() :
+                    ((HttpContent.LimitArrayPoolWriteStream)buffer).ToArray();
             }
-            catch when (LogRequestFailed(telemetryStarted))
+            catch (Exception e)
             {
-                // Unreachable as LogRequestFailed will return false
+                HandleFailure(e, telemetryStarted, response, cts, cancellationToken, timeoutTime);
                 throw;
             }
             finally
             {
-                if (HttpTelemetry.Log.IsEnabled() && telemetryStarted)
-                {
-                    if (responseContentTelemetryStarted)
-                    {
-                        HttpTelemetry.Log.ResponseContentStop();
-                    }
-
-                    HttpTelemetry.Log.RequestStop();
-                }
+                buffer?.Dispose();
```

Comment:
Any particular reason you explicitly Dispose here vs `using` in the GetString case above?

Reply:
At the time I'm sure I had a good reason :)  Changing.